### ISSUE TITLE
Debugging computed environment

### ISSUE THREAD

**Issue by clementblaise** [crossplane/crossplane#3967]
### What problem are you facing?

As stated in the design doc it should be possible to debug environment patches:
https://github.com/crossplane/crossplane/blob/edadd8c596553b290bd0d32f97c950fbcbc363d2/design/one-pager-composition-environment.md?plain=1#L100-L105

I could not see it in the logs or in the code.

### How could Crossplane help solve your problem?

It would be very useful to have access to the computed environment data as it will make troubleshooting easier.

**Comment by bobh66** [comment]
I was thinking along these same lines, both for EnvironmentConfigs and CompositionFunctions - it would be nice to have `status.environment` populated with the contents of the `environment` that was calculated for the reconciliation, and for XFN it would be nice to have `status.observed` and `status.desired` available to be able to observe what the results of the XFN are.

**Comment by bobh66** [comment]
>  it would be nice to have status.environment populated with the contents of the environment that was calculated for the reconciliation, and for XFN it would be nice to have status.observed and status.desired available to be able to observe what the results of the XFN are.

The problem with this is that the composite status gets copied to the Claim status, which would expose implementation details to the user.  So while I still think it's important to have access to the computed environment and composition function results, putting them in the status is likely not the solution.

**Comment by mateusz-lubanski-sinch** [comment]
maybe creating computed EnvironmentConfig object with `CompositeResource.metadata.uid` name would be an option?

**Comment by bobh66** [comment]
> maybe creating computed EnvironmentConfig object with CompositeResource.metadata.uid name would be an option?

That could be confusing if someone makes changes to the computed EnvironmentConfig, which then get overwritten by the next reconcile.  I still think `status.environment` on the composite would be the ideal solution, if we can prevent it from being copied to the Claim.  The status attribute name should probably be something Crossplane-specific that a user would be unlikely to define in their XRD.

**Comment by phisco** [comment]
With @bobh66 and @pedjak, we discussed briefly about this during the last sig meeting, here are a few takeaways.

Being able to see the computed environment is mostly useful for Composition authors while developing or debugging some existing Composite resource for the Composition they developed. So they should be our target.

Composition authors can already dump the whole environment to some field in the Composite resource as long as they are willing to expose the information to users, so dumping that by default wouldn't add anything, we could think of making it easier either by documenting it or by providing an automated way behind just a boolean field in Compositions, `spec.environment.addToStatus` or something similar.

Another option we discussed was to add a subcommand to the `kubectl` plugin to allow render out the `in-memory` environment for a specific Composite Resource?
Something like: `kubectl crossplane environment render <kind> <name>` which could run the same logic as Crossplane, obviously it would be on the user to match the version actually deployed. This would also be aligned with an effort @sttts is driving about gaining observability into Composition rendering.

Any thought?
